### ISSUE TITLE
I2I: Override AMP Release's least significant digit

### ISSUE THREAD

**Issue by jridgewell** [ampproject/amphtml#21952]
## Summary

Currently, the AMP release version is composed of `YYMMDDHHMMSS0` based on the commit date of the checked-out branch. That `0` can actually be `0` or `1` depending on if the current working directory is clean (`0`) or dirty (`1`).

All our releases are only on clean branches. So, all our releases have a `0` as the least significant digit. This is mainly a side-effect of a our release scripts, there's currently no way to make changes to a checkout before the process starts, so it's always a clean checkout of the desired commit.

## Design document

```diff
diff --git a/build-system/internal-version.js b/build-system/internal-version.js
index 789626fe2..2b1f06531 100644
--- a/build-system/internal-version.js
+++ b/build-system/internal-version.js
@@ -16,7 +16,7 @@
 'use strict';
 
 const argv = require('minimist')(process.argv.slice(2));
-const {gitCommitFormattedTime, gitStatusPorcelain} = require('./git');
+const {gitCommitFormattedTime} = require('./git');
 
 function getVersion() {
   if (argv.version) {
@@ -35,12 +35,8 @@ function getVersion() {
     // *Commit time is different from author time! Commit time is the time that
     // the PR was merged into master; author time is when the author ran the
     // "git commit" command.
-    const lastCommitFormattedTime = gitCommitFormattedTime();
-    if (gitStatusPorcelain()) {
-      return `${lastCommitFormattedTime}1`;
-    } else {
-      return `${lastCommitFormattedTime}0`;
-    }
+    const releaseNumber = argv.releaseNumber || 0;
+    return `${gitCommitFormattedTime()}${releaseNumber}`;
   }
 }

```

## Motivation

Sometimes it's necessary to create new AMP releases that do not change any code. I'm currently doing that right now with https://github.com/ampproject/amphtml/tree/amp-release-2019-04-23-not-a-cherry-pick, and I've done it before when our Control RTVs had the incorrect configuration applied to them. These new releases do not have code changes, but I still have to create empty commits in order to generate a new commit dates. That's because if I were to just re-run the release script, I would get the exact same commit time and the working directory would still be clean, giving the same release version.

I propose we allow the least significant digit to be configurable via a command-line flag. That way, we can easily generate new AMP releases of the same commit.

/cc @ampproject/wg-approvers
/cc @amphtml/wg-infra

**Comment by dvoytenko** [comment]
Aren't the first two significant digits indicate the version of the same release? E.g. the actual RTV is typically something like `011903262220080` where `01` is the minor release number done on the same commit.

**Comment by jridgewell** [comment]
We could do it that way, but right now we generate 4 builds of the same amp version. Specifying a new prefix might mean I have to generate `05` `06` `07` and `08`, whereas if I can specify the amp release suffix, we can take advantage of the current 4 build release process.

**Comment by dvoytenko** [comment]
Are the 4 builds actually orderly like this? I was always under impression that we simply increment each time we make a new build for any reason.

Anyhow, I don't currently see any problem with changing the last digit either. Unless release folk have some concerns...

**Comment by danielrozenberg** [comment]
As discussed offline, this seems like a reasonable, non-breaking change, and it's not like we can really rely on this `1`ness property to decide whether a release was built from an unmodified copy of the repository (it is very simple to fake) - so I have no objections to it either. Sent out #22091 to implement this, based on the implementation suggestion above

**Comment by rsimha** [comment]
To answer @dvoytenko's question:
> Are the 4 builds actually orderly like this? I was always under impression that we simply increment each time we make a new build for any reason.

The builds are indeed orderly, with `00`, `01`, `02`, and `03` always mapping to the four build flavors of `canary`, `prod`, `control`, and `rc`, in that order. In the past, we did have the odd release that incremented the 2 digit prefix, but we no longer do so.

**Comment by dvoytenko** [comment]
@rsimha between the two: prefix and suffix, what's your recommendation for the use case presented in this ITI? Will you be one of the approvers of this ITI?

**Comment by rsimha** [comment]
My opinion is that we should use the last digit, which is for all intents and purposes, not in use today.
- The first two digits have pretty much been baked in to our release process, and into the workflow around opting in to prod (by default) / canary / rc / nightly (coming soon).
- Changing the meaning of the first two digits will require spinning up four (soon to be 5) new version prefixes, will require significant changes to our serving infrastructure, and will make it significantly more difficult to add new release flavors.
- Changing the meaning of the last digit is a no-op for our serving infrastructure, and will retain the meanings of the first two digits. For example, a release version starting with `00` is guaranteed to be of the `canary` flavor, no matter what the last digit is.

I approve this change.

**Comment by jridgewell** [comment]
Would this be better if instead of being the last digit, it was the third digit of the RTV? So it would be `${PREFIX}${extra}${COMMIT_DATE}` instead of `${PREFIX}${COMMIT_DATE}${extra}`.

**Comment by rsimha** [comment]
I prefer `${PREFIX}${COMMIT_DATE}${extra}`.
- Retains the alphabetical order of releases.
- Makes searching for a version number in `v0.js` trivially easy, no matter what the least significant digit is.
- Will not require any changes to the release / serving infrastructure (other than the `--custom_version_mark` flag)

**Comment by dvoytenko** [comment]
SG. LGTM.

**Comment by danielrozenberg** [comment]
Another reason to prefer the extra digit at the last position is that it maintains the property of version numbers being monotonically increasing

**Comment by jridgewell** [comment]
I obviously approve, so I think that makes 3.

**Comment by danielrozenberg** [comment]
Was implemented in #22091
